runtime/vm/ContinuationHelpers.cpp

```diff
@@ -783,11 +783,20 @@ detachMonitorInfo(J9VMThread *currentThread, j9object_t lockObject)
 		objectMonitor = J9_INFLLOCK_OBJECT_MONITOR(lock);
 	}
 
-	J9ThreadAbstractMonitor *monitor = (J9ThreadAbstractMonitor *)objectMonitor->monitor;
-	Trc_VM_detachMonitorInfo_Detach(currentThread, currentThread->currentContinuation, objectMonitor, monitor, monitor->owner, monitor->count, currentThread->osThread);
-	monitor->owner = (J9Thread *)J9_OBJECT_MONITOR_OWNER_DETACHED;
 	Assert_VM_notNull(currentThread->currentContinuation);
-	objectMonitor->ownerContinuation = currentThread->currentContinuation;
+	J9ThreadAbstractMonitor *monitor = (J9ThreadAbstractMonitor *)objectMonitor->monitor;
+	*alreadyDetached = IS_J9_OBJECT_MONITOR_OWNER_DETACHED(monitor->owner);
+	if (*alreadyDetached) {
+		Assert_VM_true(objectMonitor->ownerContinuation == currentThread->currentContinuation);
+	} else {
+		if (NULL == objectMonitor->ownerContinuation) {
+			objectMonitor->ownerContinuation = currentThread->currentContinuation;
+			Trc_VM_detachMonitorInfo_Detach(currentThread, currentThread->currentContinuation, objectMonitor, monitor, monitor->owner, monitor->count, currentThread->osThread);
+			monitor->owner = (J9Thread *)J9_OBJECT_MONITOR_OWNER_DETACHED;
+		} else {
+			Assert_VM_unreachable();
+		}
```

Comment:
I would prefer to merge this into `else if {} ... else {}`, avoid the extra indent